core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/provision/ClusterCreationActions.java

```diff
@@ -107,11 +112,33 @@ protected Selectable createRequest(StackContext context) {
         };
     }
 
+    @Bean(name = "CLEANUP_FREEIPA_STATE")
+    public Action<?, ?> cleanupFreeIpaAction() {
+        return new AbstractStackCreationAction<>(HostMetadataSetupSuccess.class) {
+
+            @Inject
+            private InstanceMetaDataService instanceMetaDataService;
+
+            @Override
+            protected void doExecute(StackContext context, HostMetadataSetupSuccess payload, Map<Object, Object> variables) throws Exception {
+                sendEvent(context);
+            }
+
+            @Override
+            protected Selectable createRequest(StackContext context) {
+                Set<InstanceMetaData> instanceMetaData = instanceMetaDataService.findNotTerminatedForStack(context.getStack().getId());
+                Set<String> hostNames = instanceMetaData.stream().map(InstanceMetaData::getDiscoveryFQDN).collect(Collectors.toSet());
+                Set<String> ips = instanceMetaData.stream().map(InstanceMetaData::getPrivateIp).collect(Collectors.toSet());
+                return new CleanupFreeIpaEvent(context.getStack().getId(), hostNames, ips);
+            }
+        };
+    }
+
     @Bean(name = "BOOTSTRAPPING_PUBLIC_ENDPOINT_STATE")
     public Action<?, ?> bootStrappingPublicEndpointAction() {
-        return new AbstractStackCreationAction<>(HostMetadataSetupSuccess.class) {
+        return new AbstractStackCreationAction<>(StackEvent.class) {
```

Comment:
Why is this StackEvent instead of CleanupFreeIpaEvent? Is it because we don't actually care about anything in the event?